dominion/src/main/java/org/codice/dominion/interpolate/Interpolator.java

```diff
@@ -0,0 +1,436 @@
+   * @param replacements the set of replacements strings to use when interpolating
+   * @param ports the ports information already allocated from the driver
+   */
+  public Interpolator(
+      String id, String container, Map<String, String> replacements, PortFinder ports) {
+    LOGGER.debug("Interpolator({}, {}, {}, {})", id, container, replacements, ports);
+    this.id = id;
+    this.container = container;
+    this.replacements = replacements;
+    this.ports = ports;
+  }
+
+  /**
+   * Initializes a new interpolator with all the specified information already computed (typically
+   * from inside a container).
+   *
+   * @param properties properties from which to retrieve the replacement and port info
+   */
+  public Interpolator(Properties properties) {
+    LOGGER.debug("Interpolator(Properties)");
+    try {
+      this.replacements =
+          new Gson().fromJson(properties.getProperty(Interpolator.REPLACEMENTS_KEY, ""), Map.class);
+    } catch (JsonSyntaxException e) {
+      throw new InterpolationException("Unable to determined replacement information", e);
+    }
+    this.id = initFromReplacements("test.id");
+    this.container = initFromReplacements("container.name");
+    try {
+      final PortFinder portFinder =
+          new Gson().fromJson(properties.getProperty(Interpolator.PORTS_KEY, ""), PortFinder.class);
+
+      this.ports =
+          (portFinder != null)
+              ? portFinder
+              : new PortFinder(Interpolator.BASE_PORT, Interpolator.BLOCK_SIZE);
+    } catch (JsonSyntaxException e) {
+      throw new InterpolationException("Unable to determined reserved ports information", e);
+    }
+  }
+
+  /**
+   * Gets the unique identifier for the current test run.
+   *
+   * @return a unique identifier for the current test run
+   */
+  public String getUUID() {
+    return id;
+  }
+
+  /**
+   * Gets the name of the container associated with this interpolator.
+   *
+   * @return the container name
+   */
+  public String getContainer() {
+    return container;
+  }
+
+  /**
+   * Interpolates a given annotation by returning a proxy version that will automatically
+   * interpolate all <code>String</code> and <code>String[]</code> methods that are annotated with
+   * the {@link Interpolate} annotation.
+   *
+   * <p><<i>Note:</i> This method will not attempt any interpolations. Instead it will return a
+   * proxy which will delay interpolation until an annotation value marked to be interpolated is
+   * accessed. It is therefore possible that accessing a value of an interpolated annotation throws
+   * an instance of {@link InterpolationException}.
+   *
+   * @param <A> the type of annotation to be interpolated
+   * @param annotation the annotation to be interpolated
+   * @return <code>annotation</code> if no methods are requested to be interpolated or a proxy if
+   *     interpolation was requested
+   */
+  public <A extends Annotation> A interpolate(A annotation) {
```

Comment:
Since this method is particularly tricky, could you include an example in the java doc? Perhaps using the ReplaceFile annotation and explaining what effect the method would have on the target attribute? Also, should the entities in the Annotations be called methods or attributes? Methods seems incorrect to me, since they just hold data and don't actually do anything.

Reply:
I will try to add an example. From another comment, the code was moved into its own separate class so that would be much easier to do that.

As for naming, from a Java implementation standpoints, they are methods. At runtime you get an object which is an instance of your annotation interface and there you have defined methods to represent the attributes of the annotation. I can see how it is confusing, this is one of those things in Java where they implemented a new concept (i.e. Annotations) by re-using the core concept of Java (i.e. a class).